pkg/apis/ceph.rook.io/v1/status.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Based on code from https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go
+
+// SetStatusCondition sets the corresponding condition in conditions to newCondition.
+// conditions must be non-nil.
+// 1. if the condition of the specified type already exists (all fields of the existing condition are updated to
+//    newCondition, LastTransitionTime is set to now if the new status differs from the old status)
+// 2. if a condition of the specified type does not exist (LastTransitionTime is set to now() if unset, and newCondition is appended)
+func SetStatusCondition(conditions *[]Condition, newCondition Condition) {
```

Comment:
Setting conditions feels more like it should be a helper in k8sutils, rather than with the CRD types.

Reply:
I mostly copy-pasted this from the API in K8s (see the comment at the top). I wouldn't add methods that get/update/create things, but I think these helpers make sense in the API the way the k8s code does it.

Reply:
So could these helpers be moved under the `pkg/operator/k8sutil` package? I don't follow why they need to be in this package.

Reply:
It's not wrong to have code in the API. This is what the Kubernetes codebase does, and that is where I copied this from (see the comment at the  top of the file `// Based on code from https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go`). 

If this were a get/set/update method helper, I would very much agree. This is merely a (pretty dumb) low-level helper to operate on the types provided by the API, and this is IMO the most correct place to put this method. This is no different than the methods for getting resources or merging placement.

Reply:
As a brief aside but related to this topic, it is an anti-pattern to have `util` packages in the first place: https://blog.golang.org/package-names

Reply:
Sure it's not wrong to put it here, but it doesn't lend itself to code reuse across operators. If another operator wanted to use this method, would it be copied to that operator's v1 package as well? If it's not specific to the Ceph types, the k8sutil is where we've been collecting the helpers.

Reply:
> As a brief aside but related to this topic, it is an anti-pattern to have `util` packages in the first place: https://blog.golang.org/package-names

Agreed util packages are to be avoided because they are so generic. The k8sutil package could be better renamed to something like k8shelper or k8scommon, but that doesn't seem much better.

Reply:
I see the communication gap here. It **is** specific to the Ceph types because it has to handle the specifics of `LastHeartbeatTime` and the other stuff we decided to put into the type. The type is specific to Ceph, and the method cannot be made generic because of that.

For future operators, it would be good for us to suggest they use `metav1.Condition` unless they have special needs.

Reply:
Got it, this is a ceph type, makes sense.